pkg/proc/native/proc_freebsd.go

```diff
@@ -0,0 +1,358 @@
+		}
+
+		if status.StopSignal() == sys.SIGTRAP {
+			if pl_flags&sys.PL_FLAG_EXITED != 0 {
+				delete(dbp.threads, tid)
+				PtraceCont(tid, 0)
+				continue
+			} else if pl_flags&sys.PL_FLAG_BORN != 0 {
+				th, err = dbp.addThread(int(tid), false)
+				if err != nil {
+					if err == sys.ESRCH {
+						// process died while we were adding it
+						continue
+					}
+					return nil, err
+				}
+				if halt {
+					return nil, nil
+				}
+				if err = th.Continue(); err != nil {
+					if err == sys.ESRCH {
+						// thread died while we were adding it
+						delete(dbp.threads, int(tid))
+						continue
+					}
+					return nil, fmt.Errorf("could not continue new thread %d %s", tid, err)
+				}
+				continue
+			}
+		}
+
+		if th == nil {
+			continue
+		}
+
+		if (halt && status.StopSignal() == sys.SIGSTOP) || (status.StopSignal() == sys.SIGTRAP) {
+			return th, nil
+		}
+
+		// TODO(dp) alert user about unexpected signals here.
+		if err := th.resumeWithSig(int(status.StopSignal())); err != nil {
+			if err == sys.ESRCH {
+				return nil, proc.ErrProcessExited{Pid: dbp.pid}
+			}
+			return nil, err
+		}
+	}
+}
+
+// Helper function used here and in threads_freebsd.go
+// Return the status code
+func status(pid int) rune {
+	status := rune(C.find_status(C.int(pid)))
+	return status
+}
+
+// Used by stop and singleStep
+// waitFast is like wait but does not handle process-exit correctly
+func (dbp *Process) waitFast(pid int) (int, *sys.WaitStatus, error) {
+	var s sys.WaitStatus
+	wpid, err := sys.Wait4(pid, &s, 0, nil)
+	return wpid, &s, err
+}
+
+// Only used in this file
+func (dbp *Process) wait(pid, options int) (int, *sys.WaitStatus, error) {
+	var s sys.WaitStatus
+	wpid, err := sys.Wait4(pid, &s, options, nil)
+	return wpid, &s, err
+}
+
+// Only used in this file
+func (dbp *Process) exitGuard(err error) error {
+	if err != sys.ESRCH {
+		return err
+	}
+	if status(dbp.pid) == StatusZombie {
+		_, err := dbp.trapWaitInternal(-1, false)
+		return err
+	}
+
+	return err
+}
+
+// Used by ContinueOnce
+func (dbp *Process) resume() error {
+	// all threads stopped over a breakpoint are made to step over it
+	for _, thread := range dbp.threads {
+		if thread.CurrentBreakpoint.Breakpoint != nil {
+			if err := thread.StepInstruction(); err != nil {
+				return err
+			}
+			thread.CurrentBreakpoint.Clear()
+		}
+	}
+	// all threads are resumed
+	PtraceCont(dbp.pid, 0)
+	return nil
+}
+
+// Used by ContinueOnce
+// stop stops all running threads and sets breakpoints
+func (dbp *Process) stop(trapthread *Thread) (err error) {
+	if dbp.exited {
+		return &proc.ErrProcessExited{Pid: dbp.Pid()}
+	}
+	// set breakpoints on all threads
+	for _, th := range dbp.threads {
+		if th.CurrentBreakpoint.Breakpoint == nil {
+			if err := th.SetCurrentBreakpoint(); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Used by Detach
+func (dbp *Process) detach(kill bool) error {
+	return PtraceDetach(dbp.pid, 0)
```

Comment:
Must be called via our PTRACE thread (e.g. `execPtraceFunc()`).

Reply:
Doesn't work that way.